src/java/org/apache/cassandra/service/snapshot/SnapshotManager.java

```diff
@@ -80,43 +81,58 @@ public Collection<TableSnapshot> getExpiringSnapshots()
         return expiringSnapshots;
     }
 
+    public Collection<TableSnapshot> getEphemeralSnapshots()
+    {
+        return ephemeralSnapshots;
+    }
+
     public synchronized void start()
     {
-        loadSnapshots();
+        if (!snapshotsLoaded)
+            loadSnapshots();
+
         resumeSnapshotCleanup();
     }
 
     public synchronized void stop() throws InterruptedException, TimeoutException
     {
         expiringSnapshots.clear();
+        ephemeralSnapshots.clear();
         if (cleanupTaskFuture != null)
         {
             cleanupTaskFuture.cancel(false);
             cleanupTaskFuture = null;
         }
+        snapshotsLoaded = false;
     }
 
     public synchronized void addSnapshot(TableSnapshot snapshot)
     {
-        // We currently only care about expiring snapshots
-        if (snapshot.isExpiring())
+        // We currently care only about expiring and ephemeral snapshots
+        if (snapshot.isExpiring() && !expiringSnapshots.contains(snapshot))
         {
             logger.debug("Adding expiring snapshot {}", snapshot);
             expiringSnapshots.add(snapshot);
         }
+        // we are adding ephemeral snapshots only in case we are doing the initial loading
+        // as we are not interested in ephemeral snapshots in manager when node is running
+        else if (snapshot.isEphemeral() && !ephemeralSnapshots.contains(snapshot) && !snapshotsLoaded)
+        {
+            logger.debug("Adding ephemeral snapshot {}", snapshot);
+            ephemeralSnapshots.add(snapshot);
+        }
     }
 
-    @VisibleForTesting
-    protected synchronized void loadSnapshots()
+    public synchronized void loadSnapshots()
     {
         SnapshotLoader loader = new SnapshotLoader(DatabaseDescriptor.getAllDataFileLocations());
         addSnapshots(loader.loadSnapshots());
+        snapshotsLoaded = true;
     }
 
     @VisibleForTesting
     protected synchronized void addSnapshots(Collection<TableSnapshot> snapshots)
     {
-        logger.debug("Adding snapshots: {}.", Joiner.on(", ").join(snapshots.stream().map(s -> s.getId()).collect(Collectors.toList())));
```

Comment:
I moved this to addSnapshot method where I differ between expiring and ephemeral snapshot.